File: power_graph/src/power_graph/nodes/bulb.py
"""
nodes/bulb.py
──────────────────────────────────────────────────────────────────────────────
Bulb — resistive sink; no outbound pip.

States: 'off' | 'dim' | 'on' | 'blown'

  off   — no power or below threshold
  dim   — powered but low voltage (voltRatio < 0.9 or ampRatio < 1)
  on    — fully powered
  blown — permanently failed; needs replacement

Blown conditions
  volts > maxVolts (1.2× nominal) OR amps > ratedAmps × 1.5

Brightness model
  dim  → brightness = min(1, voltRatio * min(1, ampRatio)) * 0.55
  on   → brightness = min(1.0, voltRatio)
  off / blown → brightness = 0
"""

import math
from typing import Dict, List
from power_graph.node_base import NodeBase, Signal, NOMINAL_VOLTS, SpikeProfile
from power_graph.node_registry import NodeRegistry


class Bulb(NodeBase):
    """Light bulb — resistive sink."""

    type  = 'bulb'
    label = 'Light'
    group = 'Loads'
    dispatch_delay = 80

    catalog = [
        {'key': 'led-5w',   'label': 'LED 5W',     'watts': 5,   'maxVolts': 264},
        {'key': 'bulb-40w', 'label': 'Bulb 40W',   'watts': 40,  'maxVolts': 288},
        {'key': 'bulb-60w', 'label': 'Bulb 60W',   'watts': 60,  'maxVolts': 288},
        {'key': 'bulb-100w','label': 'Bulb 100W',  'watts': 100, 'maxVolts': 268},
    ]

    @classmethod
    def _default_spike(cls) -> SpikeProfile:
        return SpikeProfile(enabled=True, percent=50, duration=0.2)

    @classmethod
    def _default_pips_outbound(cls, node_id: int) -> List[Dict]:
        return []  # Bulbs are sinks

    @classmethod
    def defaults(cls, node_id: int, preset: Dict = None) -> Dict:
        if preset is None:
            preset = {}
        base = super().defaults(node_id, preset)
        watts    = preset.get('watts', 60)
        # rated amps at nominal volts
        rated_a  = watts / NOMINAL_VOLTS
        base.update({
            'watts':      watts,
            'maxVolts':   preset.get('maxVolts', 288),
            'ratedAmps':  preset.get('ratedAmps', rated_a),
            'minVolts':   preset.get('minVolts', 100),
            'blown':      False,
            'brightness': 0.0,
            'state':      'off',
        })
        return base

    @classmethod
    def config_fields(cls) -> List[str]:
        return [*super().config_fields(), 'watts', 'maxVolts', 'minVolts']

    @classmethod
    def apply(cls, panel: Dict, signal: Signal, graph):
        if panel.get('blown'):
            if panel['state'] != 'blown':
                panel['state']      = 'blown'
                panel['brightness'] = 0.0
                cls.dispatch(panel, 'bulb:blown', {'id': panel['id']})
            return

        if signal is None or signal.get('v', 0) < panel.get('minVolts', 100):
            if panel['state'] != 'off':
                cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': 'off'})
            panel['state']      = 'off'
            panel['brightness'] = 0.0
            return

        volts     = signal.get('v', 0)
        amps      = signal.get('a', 0)
        nom_volts = panel.get('maxVolts', 288) / 1.2  # nominal = maxVolts / 1.2

        # ── Blown check (overvoltage only) ────────────────────────────────────
        if volts > panel.get('maxVolts', 288):
            panel['blown']      = True
            panel['state']      = 'blown'
            panel['brightness'] = 0.0
            cls.dispatch(panel, 'bulb:blown', {'volts': volts, 'amps': amps})
            return

        volt_ratio = volts / nom_volts if nom_volts > 0 else 1.0
        amp_ratio  = (amps / panel['ratedAmps']) if panel.get('ratedAmps', 0) > 0 else 1.0
        prev_state = panel['state']

        if volt_ratio < 0.9 or amp_ratio < 0.8:
            panel['state']      = 'dim'
            panel['brightness'] = min(1.0, volt_ratio * min(1.0, amp_ratio)) * 0.55
        else:
            if prev_state == 'off':
                cls.start_spike(panel)
            panel['state']      = 'on'
            panel['brightness'] = min(1.0, volt_ratio)

        if prev_state != panel['state']:
            cls.dispatch(panel, 'state:change', {'from': prev_state, 'to': panel['state']})
        cls.throttle(panel, 'bulb:brightness', {'brightness': round(panel['brightness'], 3)})

    @classmethod
    def tick(cls, panel: Dict, dt: float, graph):
        cls.tick_spike(panel, dt)

    @classmethod
    def reset(cls, panel: Dict, graph):
        prev            = panel['state']
        panel['blown']  = False
        panel['state']  = 'off'
        panel['brightness'] = 0.0
        cls.dispatch(panel, 'state:change', {'from': prev, 'to': 'off'})


NodeRegistry.register(Bulb)

File: func-pipes/static/js/power2/nodes/bulb.js
/*
  nodes/bulb.js — Resistive Lamp (sink)
  ─────────────────────────────────────────────────────────────────────────────
  A visual power sink. Brightness grades between off, dim, and fully lit.
  No outbound pip — it consumes all power reaching it.

  Extended state
  ──────────────
  watts       number  — rated wattage (determines amp draw)
  maxVolts    number  — above this the bulb blows (overvoltage)
  maxAmps     number  — above this the bulb blows (overcurrent)
  brightness  number  — 0.0–1.0 (used for CSS / animation)
  blown       bool    — destroyed by overvoltage or overcurrent; requires manual reset

  States: 'off' | 'dim' | 'on' | 'blown'
*/

class Bulb extends NodeBase {

    static type  = 'bulb'
    static label = 'Bulb'
    static group = 'Light'

    static _defaultSpike() {
        return { enabled: true, percent: 25, duration: 0.73 }
    }

    static catalog = [
        { key: 'led-5w',    label: 'LED  5W',    watts: 5   },
        { key: 'bulb-40w',  label: 'Bulb  40W',  watts: 40  },
        { key: 'bulb-60w',  label: 'Bulb  60W',  watts: 60  },
        { key: 'bulb-100w', label: 'Bulb 100W',  watts: 100 },
    ]

    static defaults(id, preset = {}) {
        const watts = preset.watts ?? 60
        return {
            ...super.defaults(id, preset),
            watts,
            maxVolts:   preset.maxVolts ?? (preset.volts ? preset.volts * 1.2 : 288),
            maxAmps:    preset.maxAmps  ?? +((watts / NOMINAL_VOLTS) * 2).toFixed(3),
            brightness: 0,
            blown:      false,
            spike:      preset.spike ? { ...preset.spike } : { ...this._defaultSpike() },
            _spikeWatts: 0,
            // Bulbs are sinks — no outbound pip
            pipsOutbound: [],
        }
    }

    static configFields() {
        return [...super.configFields(), 'watts', 'maxVolts', 'maxAmps', 'spike']
    }

    static apply(panel, signal, graph) {
        if (panel.blown) return   // blown = open circuit until manually reset

        if (!signal || signal.v <= 0 || signal.a <= 0) {
            const prev = panel.state
            panel.state      = 'off'
            panel.brightness = 0
            if (prev !== 'off') Bulb.dispatch(panel, 'state:change', { from: prev, to: 'off' })
            graph.emit(panel, null)
            return
        }

        if (signal.v > panel.maxVolts) {
            const prev = panel.state
            panel.blown      = true
            panel.state      = 'blown'
            panel.brightness = 0
            Bulb.dispatch(panel, 'bulb:blown', { reason: 'overvoltage', volts: signal.v, maxVolts: panel.maxVolts })
            Bulb.dispatch(panel, 'state:change', { from: prev, to: 'blown' })
            graph.emit(panel, null)
            return
        }

        if (signal.a > panel.maxAmps) {
            const prev = panel.state
            panel.blown      = true
            panel.state      = 'blown'
            panel.brightness = 0
            Bulb.dispatch(panel, 'bulb:blown', { reason: 'overcurrent', amps: signal.a, maxAmps: panel.maxAmps })
            Bulb.dispatch(panel, 'state:change', { from: prev, to: 'blown' })
            graph.emit(panel, null)
            return
        }

        const drawAmps  = panel.watts / NOMINAL_VOLTS
        const voltRatio = signal.v / NOMINAL_VOLTS
        const ampRatio  = signal.a / drawAmps

        const prev      = panel.state
        const prevBrightness = panel.brightness

        if (signal.v < 180 || signal.a < drawAmps * 0.05) {
            panel.state      = 'off'
            panel.brightness = 0
        } else if (voltRatio < 0.9 || ampRatio < 1) {
            panel.state      = 'dim'
            panel.brightness = Math.min(1, voltRatio * Math.min(1, ampRatio)) * 0.55
            if (prev !== 'dim') Bulb.dispatch(panel, 'bulb:brownout', { volts: signal.v, amps: signal.a })
        } else {
            panel.state      = 'on'
            panel.brightness = Math.min(1.0, voltRatio)
            if (prev !== 'on') NodeBase.startSpike(panel)
        }

        if (panel.state !== prev)
            Bulb.dispatch(panel, 'state:change', { from: prev, to: panel.state })
        if (Math.abs(panel.brightness - prevBrightness) >= 0.05)
            Bulb.dispatch(panel, 'bulb:brightness', { brightness: +panel.brightness.toFixed(2) })

        graph.emit(panel, null)   // sink — nothing forwarded
    }

    static tick(panel, dt, graph) {
        const wasSpiking = NodeBase.tickSpike(panel, dt)
        // Track inflated watt draw so computeGenDraw picks it up during the spike.
        panel._spikeWatts = (panel.state === 'on' || panel.state === 'dim')
            ? panel.watts * NodeBase.spikeMultiplier(panel)
            : 0
        if (wasSpiking && panel.state !== 'off' && panel.state !== 'blown' && panel.signal)
            Bulb.apply(panel, panel.signal, graph)
    }

    static reset(panel, graph) {
        panel.blown      = false
        panel.brightness = 0
        panel.powerSources = {}
        Bulb.dispatch(panel, 'bulb:reset', {})
        super.reset(panel, graph)
    }
}

NodeRegistry.register(Bulb)
